Math/NumberTheory/DirichletCharacters.hs

```diff
@@ -0,0 +1,589 @@
+-- |
+-- Module:      Math.NumberTheory.DirichletCharacters
+-- Copyright:   (c) 2018 Bhavik Mehta
+-- Licence:     MIT
+-- Maintainer:  Bhavik Mehta <bhavikmehta8@gmail.com>
+-- Stability:   Provisional
+-- Portability: Non-portable (GHC extensions)
+--
+-- Implementation and enumeration of Dirichlet characters.
+--
+
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE CPP #-}
+
+module Math.NumberTheory.DirichletCharacters
+  (
+  -- * Roots of unity
+    RootOfUnity
+  -- ** Conversions
+  , toRootOfUnity
+  , fromRootOfUnity
+  , toComplex
+  -- * An absorbing semigroup
+  , OrZero, pattern Zero, pattern NonZero
+  , asNumber
```

Comment:
I would probably do not export this function. It is just `maybe 0`, not worth of polluting global namespace with a rather ambiguous name.

Reply:
It's `maybe 0` but with the newtype abstraction `Ap` in the way, so this would be useful to export. I'll happily take better name suggestions.

Reply:
Ah, I am terrible with names. Probably `orZeroToNum` will do, if you still prefer it to `coerce (maybe 0)`.